metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/writer/CachedFieldNameConverterFactory.java

```diff
@@ -0,0 +1,155 @@
+ * from the cache.
+ *
+ * <p>Defining a shorter expiration interval allows config changes to be recognized more
+ * quickly, but also can impact performance negatively.
+ */
+public class CachedFieldNameConverterFactory implements FieldNameConverterFactory {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  /**
+   * A cache that contains a {@link FieldNameConverter} for each sensor type.
+   *
+   * A user can alter the {@link FieldNameConverter} for a given sensor at any time
+   * by altering the Indexing configuration.  The actual {@link FieldNameConverter}
+   * in use for a given sensor will only change once the original converter has
+   * expired from the cache.
+   */
+  private Cache<String, FieldNameConverter> fieldNameConverters;
+
+  /**
+   * Creates a {@link CachedFieldNameConverterFactory}.
+   *
+   * @param expires The duration before {@link FieldNameConverter}s are expired.
+   * @param expiresUnits The units before {@link FieldNameConverter}s are expired.
+   */
+  public CachedFieldNameConverterFactory(int expires, TimeUnit expiresUnits) {
+
+    fieldNameConverters = createFieldNameConverterCache(expires, expiresUnits);
+  }
+
+  /**
+   * Creates a {@link CachedFieldNameConverterFactory} where the cache expires after 5 minutes.
+   */
+  public CachedFieldNameConverterFactory() {
+
+    this(5, TimeUnit.MINUTES);
+  }
+
+  /**
+   * Creates a {@link CachedFieldNameConverterFactory} using the given cache.  This should only
+   * be used for testing.
+   *
+   * @param fieldNameConverters A {@link Cache} containing {@link FieldNameConverter}s.
+   */
+  public CachedFieldNameConverterFactory(Cache<String, FieldNameConverter> fieldNameConverters) {
+
+    this.fieldNameConverters = fieldNameConverters;
+  }
+
+  /**
+   * Creates a cache of {@link FieldNameConverter}s, one for each source type.
+   *
+   * @return A cache of {@link FieldNameConverter}s.
+   */
+  private Cache<String, FieldNameConverter> createFieldNameConverterCache(int expire, TimeUnit expireUnits) {
+
+    return Caffeine
+            .newBuilder()
+            .expireAfterWrite(expire, expireUnits)
+            .build();
+  }
+
+  /**
+   * Create a new {@link FieldNameConverter}.
+   *
+   * @param sensorType The type of sensor.
+   * @param config The writer configuration.
+   * @return
+   */
+  @Override
+  public FieldNameConverter create(String sensorType, WriterConfiguration config) {
+
+    return fieldNameConverters.get(sensorType, (s) -> createInstance(sensorType, config));
+  }
+
+  /**
+   * Create a new {@link FieldNameConverter}.
+   *
+   * @param sensorType The type of sensor.
+   * @param config The writer configuration.
+   * @return
+   */
+  private FieldNameConverter createInstance(String sensorType, WriterConfiguration config) {
+
+    // default to the 'DEDOT' field name converter to maintain backwards compatibility
```

Comment:
This looks interesting, but one bit of functionality has changed as part of doing this PR.  Currently, we specify the field converter at the writer implementation level, so:
* ES uses dedot
* HDFS uses noop
* solr uses noop

By doing this, we're actually not maintaining backwards compatibility, we're changing the behavior for the HDFS writer and Solr to dedot.  What I'd suggest doing is adding a method to the `BulkMessageWriter` interface like so:
```
default FieldNameConverter defaultFieldNameConverter() {
  return FieldNameConverters.NOOP.get();
}
```
and in ElasticsearchWriter specify DEDOT as the default.  Also, here, you probably want to pass in the default field name converter if unspecified as a 3rd argument.

This would allow us to maintain backwards compatibility and enable users to override.

Reply:
Remember that only the `ElasticsearchWriter` does field name conversion.  HDFS and Solr do not do field name conversion and so are backwards compatible.  

In the README, I think I specified that this field is only applicable for 'elasticsearch'.

In the future, we could certainly role this functionality out to the other Writers, but right now its not there. AFAIK

Reply:
OH! yes, you're right.  Sorry for jumping to conclusions, I assumed you generalized this to the bulk message writer level, rather than keeping it at the ES writer level.  My bad, I retract the objection.